11_FFTBloom/app_resources/kernel_fft_first_axis.hlsl

```diff
@@ -0,0 +1,154 @@
+
+struct SharedMemoryAccessor 
+{
+	void set(uint32_t idx, uint32_t value) 
+	{
+		sharedmem[idx] = value;
+	}
+	
+	void get(uint32_t idx, NBL_REF_ARG(uint32_t) value) 
+	{
+		value = sharedmem[idx];
+	}
+
+	void workgroupExecutionAndMemoryBarrier() 
+	{
+		glsl::barrier();
+    }
+
+};
+
+// ---------------------------------------------------- Utils ---------------------------------------------------------
+
+// After first FFT we only store half of a column, so the offset per column is half the image side length
+uint64_t colMajorOffset(uint32_t x, uint32_t y)
+{
+	return x * (IMAGE_SIDE_LENGTH / 2) | y;
+}
+
+// Each channel after first FFT will be stored as half the image (cut along the x axis) in col-major order, and the whole size of the image is N^2, 
+// for N = _NBL_HLSL_WORKGROUP_SIZE_ * ELEMENTS_PER_THREAD
+uint64_t getChannelStartAddress(uint32_t channel)
+{
+	return pushConstants.colMajorBufferAddress + channel * IMAGE_SIDE_LENGTH * IMAGE_SIDE_LENGTH / 2 * sizeof(complex_t<scalar_t>);
+}
+
+// -------------------------------------------- FIRST AXIS FFT ------------------------------------------------------------------
+
+// Each Workgroup computes the FFT along two consecutive vertical scanlines (fixed x for the whole Workgroup) so we use `2 * gl_WorkGroupID().x, 2 * gl_WorkGroupID().x + 1` 
+// to get the x coordinates for each of the consecutive lines
+// Since the output images (one per channel) are square of size IMAGE_SIDE_LENGTH (defined above) we will be launching half that amount of workgroups
+
+struct PreloadedFirstAxisAccessor : PreloadedAccessorBase<ELEMENTS_PER_THREAD, _NBL_HLSL_WORKGROUP_SIZE_, scalar_t> 
+{
+	
+	void preload(uint32_t channel)
+	{
+		float32_t2 inputImageSize;
+		texture.GetDimensions(inputImageSize.x, inputImageSize.y);
+		float32_t2 normalizedCoordsFirstLine, normalizedCoordsSecondLine;
+		normalizedCoordsFirstLine.x = (float32_t(2 * glsl::gl_WorkGroupID().x)+0.5f)/(inputImageSize.x*KERNEL_SCALE);
+		normalizedCoordsSecondLine.x = (float32_t(2 * glsl::gl_WorkGroupID().x + 1)+0.5f)/(inputImageSize.x*KERNEL_SCALE);
+
+		for (uint32_t localElementIndex = 0; localElementIndex < ELEMENTS_PER_THREAD; localElementIndex++)
+		{
+			// Index computation here is easier than FFT since the stride is fixed at _NBL_HLSL_WORKGROUP_SIZE_
+			const uint32_t index = _NBL_HLSL_WORKGROUP_SIZE_ * localElementIndex | workgroup::SubgroupContiguousIndex();
+			normalizedCoordsFirstLine.y = (float32_t(index) + 0.5f) / (inputImageSize.y * KERNEL_SCALE);
+			normalizedCoordsSecondLine.y = normalizedCoordsFirstLine.y;
+			preloaded[localElementIndex].real(scalar_t(texture.SampleLevel(samplerState, normalizedCoordsFirstLine + promote<float32_t2, float32_t>(0.5 - 0.5 / KERNEL_SCALE), 0)[channel]));
+			preloaded[localElementIndex].imag(scalar_t(texture.SampleLevel(samplerState, normalizedCoordsSecondLine + promote<float32_t2, float32_t>(0.5 - 0.5/KERNEL_SCALE), 0)[channel]));
```

Comment:
why are you using a sampler and normalized coords here? its only the input image we want to use wrap modes for, the kernel image must/should be PoT sized

Reply:
USe `Texture2DArray::Load` method instead of `SampleLevel` here https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-to-load

Reply:
KERNEL SCALE is a super rare use-case, in the future I'd rather use a mip-map (scale down in PoT increments) if minification is required

Reply:
This is what old example does to achieve a minified version of the kernel (centered in the middle of an image with the same size as the original kernel, with 0 padding all around) when `bloomScale < 1`, idk what to do otherwise

Reply:
your choice either keep the scale and expose it or just fix it to equal 1.f